cmake/Components.cmake

```diff
@@ -0,0 +1,145 @@
+function(add_component_include_directories component)
+  set(target "${component_prefix}${component}")
+  set_property(TARGET "${target}" APPEND
+    PROPERTY "${component_prefix}INCLUDE_DIRS" ${ARGN})
+endfunction()
+
+#==============================================================================
+function(add_component_dependencies component)
+  set(dependency_components ${ARGN})
+
+  is_component(is_valid "${component}")
+  if(NOT ${is_valid})
+    message(FATAL_ERROR
+      "Target '${component}' is not a component of ${PROJECT_NAME}.")
+  endif()
+
+  set(target "${component_prefix}${component}")
+
+  foreach(dependency_component ${dependency_components})
+    is_component(is_valid "${dependency_component}")
+    if(NOT ${is_valid})
+      message(FATAL_ERROR
+        "Target '${dependency_component}' is not a component of ${PROJECT_NAME}.")
+    endif()
+
+    set(dependency_target "${component_prefix}${dependency_component}")
+    add_dependencies("${target}" "${dependency_target}")
+  endforeach()
+
+  set_property(TARGET "${target}" APPEND
+    PROPERTY "${component_prefix}DEPENDENCIES" ${dependency_components})
+endfunction()
+
+#==============================================================================
+function(add_component_targets component)
+  set(dependency_targets ${ARGN})
+
+  is_component(is_valid "${component}")
+  if(NOT ${is_valid})
+    message(FATAL_ERROR
+      "Target '${component}' is not a component of ${PROJECT_NAME}.")
+  endif()
+
+  set(target "${component_prefix}${component}")
+  add_dependencies("${target}" ${ARGN})
+
+  foreach(dependency_target ${dependency_targets})
+    if(NOT TARGET "${dependency_target}")
+      message(FATAL_ERROR "Target '${dependency_target}' does not exist.")
+    endif()
+
+    get_property(dependency_type TARGET "${dependency_target}" PROPERTY TYPE)
+    if(NOT ("${dependency_type}" STREQUAL STATIC_LIBRARY
+         OR "${dependency_type}" STREQUAL SHARED_LIBRARY))
+      message(FATAL_ERROR
+        "Target '${dependency_target}' has unsupported type"
+        " '${dependency_type}'. Only 'STATIC_LIBRARY' and 'SHARED_LIBRARY'"
+        " are supported.")
+    endif()
+
+    install(TARGETS "${dependency_target}"
+      EXPORT "${component_prefix}${component}"
+      ARCHIVE DESTINATION "${LIBRARY_INSTALL_DIR}"
+      LIBRARY DESTINATION "${LIBRARY_INSTALL_DIR}"
+    )
+  endforeach()
+
+  set_property(TARGET "${target}" APPEND
+    PROPERTY "${component_prefix}LIBRARIES" ${dependency_targets})
+endfunction()
+
+#==============================================================================
+function(install_component_exports)
+  get_property(components GLOBAL PROPERTY "${global_prefix}COMPONENTS")
+
+  set(output_prefix "${CMAKE_CURRENT_BINARY_DIR}/${CONFIG_INSTALL_DIR}")
+
+  foreach(component ${components})
+    set(target "${component_prefix}${component}")
+
+    # TODO: Replace this manual generation with a configure_file.
+    set(output_path
```

Comment:
This is the only part that feels not great. I'm sure it works, but writing directly to a file from CMake seems like an abuse. What would the config file look like that you can reuse multiple times here?